### PR TITLE
feat: get token info from web3 network

### PR DESCRIPTION
**Pull request by remi-guan**
Current progress:

- [x] Display logs when a user sent a POST request with an invalid Body object.
- [x] Get token info from the web3 network.
  > TODO: support more `chainId`; validate supported `chainId`
- [x] Verify input param contract address.
- [x] Change the new API endpoints to the RESTful format.
- [ ] `amount` column of Price support BigNumber as input.
  > TODO: API and Database have to support a very large `amount` number as the input with BigNumber.js, or use string instead.

### CODE REVIEW

File: src/api/blockchains/tokens/pipes/address-validation.pipe.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    const { chainId, address } = value;
+    let isContract: boolean;
+
+    try {
+      isContract = await new Web3Provider(
+        this.configService,
+        Number(chainId),
+      ).isContract(address);
+    } catch (error) {
+      throw new BadRequestException('Address is not a valid Ethereum address');
```
Comment (w3itch) [inline src/api/blockchains/tokens/pipes/address-validation.pipe.ts:19]: log error

File: src/api/games/game-projects.controller.ts [hunk 2]
```diff
@@ -253,14 +248,9 @@ export class GameProjectsController {
...
-    const target = await this.gamesService.findOne(id);
-
-    if (user.username !== target.username) {
-      throw new ForbiddenException(
-        "You don't have permission to delete the file of this project",
-      );
-    }
+    await this.gamesService.verifyOwner(id, user);
 
+    const target = await this.gamesService.findOne(id);
```
Comment (w3itch) [inline src/api/games/game-projects.controller.ts:253]: extract a method of gamesService

File: src/api/prices/prices.module.ts [hunk 3]
```diff
@@ -16,7 +17,7 @@ import { PricesService } from './prices.service';
     TypeOrmModule.forFeature([Game, Price, Token]),
```
Comment (w3itch) [inline src/api/prices/prices.module.ts]: import GamesModule & TokensModule Instead

File: src/constants/web3-providers.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export const web3Providers: Record<number, { url: string; token: string }> = {
```
Comment (w3itch) [inline src/constants/web3-providers.ts]: Move to blockchains/tokens/web3.providers.ts

File: src/entities/Game.entity.ts [hunk 5]
```diff
@@ -146,15 +165,17 @@ export class Game extends BaseEntity {
...
+    enum: Community,
+    default: Community.DISQUS,
   })
   @Column({ default: Community.DISQUS })
   @IsEnum(Community)
   @IsNotEmpty()
   community: Community;
 
   @ApiProperty({
-    description: 'The category that best describes this game',
```
Comment (w3itch) [inline src/entities/Game.entity.ts:157]: Why did you remove the description？

File: src/api/blockchains/tokens/pipes/address-validation.pipe.ts (hunk 1 above)
Comment (remi-guan) [reply src/api/blockchains/tokens/pipes/address-validation.pipe.ts:17]: > Error may be thrown due to infura configuration etc., exception description is not accurate

I've moved the `await new Web3Provider(...)` statement outside of the `try...catch` blocks, if it throws an error it will not be caught.

**Review by w3itch — COMMENTED** [review]
- Change typeof `amount` in Price
- Restore API property description
- Refactoring controller, business logic should be placed in service

File: src/api/prices/prices.module.ts (hunk 3 above)
Comment (remi-guan) [reply src/api/prices/prices.module.ts]: Marked as Resolved due to outdated.

File: src/entities/Game.entity.ts (hunk 5 above)
Comment (remi-guan) [reply src/entities/Game.entity.ts:157]: Yeah, it was by accident. I've copied them back.

File: src/constants/web3-providers.ts (hunk 4 above)
Comment (remi-guan) [reply src/constants/web3-providers.ts]: Marked as Resolved due to outdated.

File: src/api/games/game-projects.controller.ts (hunk 2 above)
Comment (remi-guan) [reply src/api/games/game-projects.controller.ts]: Yeah I've implemented a new service called `game.logic.service.ts`, check out my new updates.

File: src/api/games/games.module.ts [hunk 6]
```diff
@@ -3,16 +3,18 @@ import { TypeOrmModule } from '@nestjs/typeorm';
...
 import { EasyRpgGamesService } from './easy-rpg.games.service';
 import { GameProjectsController } from './game-projects.controller';
+import { GamesBaseService } from './games.base.service';
 import { GamesController } from './games.controller';
-import { GamesService } from './games.service';
+import { GamesLogicService } from './games.logic.service';
 
 @Module({
-  imports: [TagsModule, TypeOrmModule.forFeature([Game, Tag])],
+  imports: [TagsModule, PricesModule, TypeOrmModule.forFeature([Game, Tag])],
```
Comment (w3itch) [inline src/api/games/games.module.ts:15]: is `TypeOrmModule.forFature([Tag])` necessary ?

File: src/entities/Price.entity.ts [hunk 7]
```diff
@@ -13,10 +13,9 @@ export class Price extends BaseEntity {
   @IsPositive()
   chainId: number;
 
-  @Column({ type: 'decimal' })
-  @IsInt()
-  @IsPositive()
```
Comment (w3itch) [inline src/entities/Price.entity.ts:18]: `amount` should be positive. Find an alternative

**Review by w3itch — APPROVED** [review]
